galasa-managers-parent/galasa-managers-comms-parent/dev.galasa.mq.manager/src/main/java/dev/galasa/mq/internal/MQManagerImpl.java

```diff
@@ -0,0 +1,258 @@
+import dev.galasa.framework.spi.AnnotatedField;
+import dev.galasa.framework.spi.ConfigurationPropertyStoreException;
+import dev.galasa.framework.spi.GenerateAnnotatedField;
+import dev.galasa.framework.spi.IFramework;
+import dev.galasa.framework.spi.IManager;
+import dev.galasa.framework.spi.ResourceUnavailableException;
+import dev.galasa.framework.spi.creds.CredentialsException;
+import dev.galasa.framework.spi.creds.ICredentialsService;
+import dev.galasa.framework.spi.language.GalasaMethod;
+import dev.galasa.framework.spi.language.GalasaTest;
+import dev.galasa.mq.IMessageQueue;
+import dev.galasa.mq.IMessageQueueManager;
+import dev.galasa.mq.MqManagerException;
+import dev.galasa.mq.MqManagerField;
+import dev.galasa.mq.Queue;
+import dev.galasa.mq.QueueManager;
+import dev.galasa.mq.internal.properties.InstanceChannelName;
+import dev.galasa.mq.internal.properties.InstanceCredentials;
+import dev.galasa.mq.internal.properties.InstanceForTag;
+import dev.galasa.mq.internal.properties.InstanceHost;
+import dev.galasa.mq.internal.properties.MqPropertiesSingleton;
+import dev.galasa.mq.internal.properties.QueueNameForTag;
+import dev.galasa.mq.internal.properties.InstanceName;
+import dev.galasa.mq.internal.properties.InstancePort;
+
+@Component(service = { IManager.class })
+public class MQManagerImpl extends AbstractManager {
+
+    private static final Log  logger = LogFactory.getLog(MQManagerImpl.class);
+    private static final String NAMESPACE   = "mq";
+    
+    private Path storedArtifactsRoot;
+    private ICredentialsService credentialService;
+    
+    private HashMap<String,MessageQueueManagerImpl> queueManagers = new HashMap<>();
+    private List<MessageQueueImpl> queues = new ArrayList<>();
+    
+    private String currentMethod = new String();
+
+    @GenerateAnnotatedField(annotation = Queue.class)
+    public IMessageQueue generateMessageQueue(Field field, List<Annotation> annotations) throws MqManagerException {
+    	Queue annotation = field.getAnnotation(Queue.class);
+    	
+    	//check that we have provisioned the qmgr for this queue
+    	String qmgrTag = annotation.queueMgrTag();
+    	if(!queueManagers.containsKey(qmgrTag)) {
+    		throw new MqManagerException("Unable to provision queue: " + annotation.name() + " no QMGR found tagged: " + qmgrTag);
+    	}
+    	
+    	//check if the name is specified in the queue or in the CPS
+    	String queueName = annotation.name();
+    	String queueTag = annotation.tag();
+    	
+    	if(queueName.isEmpty() && queueTag.isEmpty())
+    		throw new MqManagerException("Either name or tag must be specified in @Queue annotation");
+    	
+    	if(!queueName.isEmpty() && !queueTag.isEmpty())
+    		throw new MqManagerException("Both name and tag are specified in @Queue annotation, these are mutually exclusive");
+    	
+    	String name = queueName;
+    	if(queueTag.length() > 0){
+    		name = QueueNameForTag.get(queueTag);
```

Comment:
Should check here if the TAGed queue has already been provisioned.   Just in case there are 2 fields with the same tag.

Reply:
ie,  exists already in this.queues

Reply:
Problem is that we have no way at the moment of uniquely identifying a queue.  Some may be named, some may be tagged, in all cases the tag gets translated to a name.  However that name is scoped to the QMGR.  to check this we would have to deep compare queue objects.  This could be done but I feel it is a lot of extra work to prevent an extra object being created.  Happy to discuss it though

Reply:
one method might be to create a unique name for each queue of the form <qmgr:queuename> and then switch the queues from being held in a list to being held in a hash map.  Then I could perform the check

Reply:
@Michael-Baylis  I've added an implementation to fix this now

Reply:
incorrect implementation I am afraid

Reply:
The key must be the TAG.     2 queue annotations with the same TAG must point to the same object.